### PR TITLE
[REVIEW] Testing code cleanup & some utility code include reordering

### PR DESCRIPTION
**Pull request by eyalroz**
Some housekeeping/cleanup of the code under `cpp/tests` (which also splits off some of the changes in PR #902 ):

* Reversed include orders - see issue #681; this exposed some missing includes, so:
* Added missing includes, especially for standard library headers, to some test programs
* Moved some non-inlined, non-template test utility code from headers into actual source files (so that it doesn't get defined twice)
* Minor changes to the column printing code: Promoting int8_t's to int's for printing; supporting a minimum width parameter when printing out columns.

No changes to cudf functionality (nor to which tests are run).

### CODE REVIEW

**Comment by eyalroz** [comment]
I wonder whom I should ask to review this... @kkraus14 : Would you be the right person?

**Comment by jrhemstad** [comment]
I'll review it. There looks to be some conflicting/redundant work with my own test utility cleanup work: https://github.com/rapidsai/cudf/pull/694 but there looks to be plenty of other sorely needed updates orthogonal to my own.

**Comment by eyalroz** [comment]
@jrhemstad : Yes, there are a few conflicts - I think mostly with the `print_gdf_column` code. But I'm guessing we were going in generally the same direction so it shouldn't be hard to merge.

**Comment by jrhemstad** [comment]
@eyalroz code you resolve conflicts with this PR so that I can review it? We'd like to get it into the 0.6 release if possible.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by eyalroz** [comment]
@jrhemstad : Conflicts resolved.

**Comment by kkraus14** [comment]
add to whitelist

File: cpp/tests/types/dispatcher_test.cu [hunk 1]
```diff
@@ -13,139 +13,126 @@
...
-TYPED_TEST_CASE(TypedDispatcherTest, TestTypes);
+namespace{ 
 
-namespace {
-template <typename ExpectedType>
-struct type_tester {
-  template <typename DispatchedType>
-  bool operator()() {
-    return std::is_same<ExpectedType, DispatchedType>::value;
+struct test_functor {
```
Comment (jrhemstad) [inline cpp/tests/types/dispatcher_test.cu]: This conflict resolution is incorrect, it overwrote the new `type_tester` functor.

File: cpp/tests/types/dispatcher_test.cu [hunk 2]
```diff
@@ -13,139 +13,126 @@
...
 
-// Ensure that the type_to_gdf_dtype trait maps to the correct gdf_dtype
-TYPED_TEST(TypedDispatcherTest, TraitsTest) {
-  EXPECT_TRUE(cudf::type_dispatcher(cudf::gdf_dtype_of<TypeParam>(),
-                                    type_tester<TypeParam>{}));
+__global__ 
+void dispatch_test_kernel(gdf_dtype type, bool * d_result)
+{
+  if(0 == threadIdx.x + blockIdx.x * blockDim.x)
+    *d_result = cudf::type_dispatcher(type, test_functor{}, type);
```
Comment (jrhemstad) [inline cpp/tests/types/dispatcher_test.cu]: Incorrect merge resolution. Overwrote `type_tester`.

File: cpp/tests/types/dispatcher_test.cu (hunk 1 above)
Comment (eyalroz) [reply cpp/tests/types/dispatcher_test.cu]: >  I'll review in-depth once the merge conflicts have been correctly resolved.

Are the places you specifically marked here the only incorrect merges?

git did this automagically and I didn't even touch `dispatcher_test.cu` manually. Anyway, I'll have another look at the entire diff with branch-0.6.

File: cpp/tests/utilities/cudf_test_utils.cu [hunk 3]
```diff
@@ -1,6 +1,10 @@
 /*
  * Copyright (c) 2018, NVIDIA CORPORATION.
  *
+ * Copyright 2018 BlazingDB, Inc.
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cu]: If this is a new Blazing copyright, shouldn't it be 2019?

File: cpp/tests/utilities/cudf_test_utils.cu [hunk 4]
```diff
@@ -1,6 +1,10 @@
...
+ *     Copyright 2019 Eyal Rozenberg <eyalroz@blazingdb.com>
+ *
+ *
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cu]:
```suggestion
```

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 5]
```diff
@@ -13,37 +13,97 @@
...
+// We use this single character to represent a gdf_column element being null
+constexpr const char null_representative = '@';
 
 // Type for a unique_ptr to a gdf_column with a custom deleter
 // Custom deleter is defined at construction
 using gdf_col_pointer = typename std::unique_ptr<gdf_column, 
                                                  std::function<void(gdf_column*)>>;
 
+template <typename Element>
+void print_typed_column(
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]: Documentation for this function.

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 6]
```diff
@@ -13,37 +13,97 @@
...
+      else {
+          std::cout << null_representative;
+      }
+      if (i + 1 < num_rows) { std::cout << ' '; }
+  }
+  std::cout << std::endl;
+}
+
+template <typename Element>
+inline void print_typed_column(const gdf_column& column, unsigned min_printing_width = 1)
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]: Documentation.

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 7]
```diff
@@ -56,11 +116,11 @@ void print_valid_data(const gdf_valid_type *validity_mask,
 
 /* --------------------------------------------------------------------------*/
 /**
- * @brief  Creates a unique_ptr that wraps a gdf_column structure intialized with a host vector
+ * @Synopsis  Creates a unique_ptr that wraps a gdf_column structure intialized with a host vector
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]: Please use `@brief` as was used before (and everywhere else).

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 8]
```diff
@@ -56,11 +116,11 @@ void print_valid_data(const gdf_valid_type *validity_mask,
...
  *
- * @param host_vector The host vector whose data is used to initialize the gdf_column
+ * @Param host_vector The host vector whose data is used to initialize the gdf_column
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]:
```suggestion
 * @param host_vector The host vector whose data is used to initialize the gdf_column
```

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 9]
```diff
@@ -56,11 +116,11 @@ void print_valid_data(const gdf_valid_type *validity_mask,
...
  *
- * @returns A unique_ptr wrapping the new gdf_column
+ * @Returns A unique_ptr wrapping the new gdf_column
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]:
```suggestion
 * @returns A unique_ptr wrapping the new gdf_column
```

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 10]
```diff
@@ -70,7 +130,9 @@ gdf_col_pointer create_gdf_column(std::vector<ColumnType> const & host_vector,
   // Get the corresponding gdf_dtype for the ColumnType
   gdf_dtype gdf_col_type{cudf::gdf_dtype_of<ColumnType>()};
 
-  EXPECT_TRUE(GDF_invalid != gdf_col_type);
+// Don't use gtest functionality within what is non-gtest code
+
+// EXPECT_TRUE(GDF_invalid != gdf_col_type);
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]: Don't leave commented out code.  Suggest you change this to `CUDF_EXPECTS()`.

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 11]
```diff
@@ -223,7 +286,8 @@ std::vector<gdf_col_pointer> initialize_gdf_columns(
  * @return bool Whether or not the columns are equal
  * ---------------------------------------------------------------------------**/
 template <typename T>
-bool gdf_equal_columns(gdf_column* left, gdf_column* right) {
+bool gdf_equal_columns(gdf_column* left, gdf_column* right)
+{
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]: Please don't make changes that *only* move brackets around. Bracket placement is not dictated by a coding standard so you are only wasting reviewers' time.

File: cpp/tests/utilities/cudf_test_utils.cuh [hunk 12]
```diff
@@ -167,6 +229,7 @@ convert_tuple_to_gdf_columns(std::vector<gdf_col_pointer> &gdf_columns,std::tupl
 }
 
 // Converts a tuple of host vectors into a vector of gdf_columns
+
```
Comment (harrism) [inline cpp/tests/utilities/cudf_test_utils.cuh]:
```suggestion
```

File: cpp/tests/utilities/cudf_test_utils.cuh (hunk 7 above)
Comment (eyalroz) [reply cpp/tests/utilities/cudf_test_utils.cuh]: This was a mistake, I didn't mean for that change to occur.

File: cpp/tests/utilities/cudf_test_utils.cuh (hunk 11 above)
Comment (eyalroz) [reply cpp/tests/utilities/cudf_test_utils.cuh]: Ok, but - this _was_ a no-functionality-changes commit, and indentation fixes seem like they go well with include reordering - very minor stuff..

**Comment by eyalroz** [comment]
@harrism : I think I've addressed all the issues you've brought up.
@jrhemstad : Ditto, a while ago already IIANM.

**Comment by harrism** [comment]
@eyalroz please merge the latest branch-0.6 into your branch, resolve conflicts, and push again.
@jrhemstad please re-review

**Comment by eyalroz** [comment]
@harrism : Thanks.

> @eyalroz please merge the latest branch-0.6 into your branch, resolve conflicts, and push again.

@jrhemstad  :  Done. A bit exasperated after the third rebase of this kind for a house-keeping PR ... if you  merge this today you'll save me a fourth.

**Comment by kkraus14** [comment]
rerun tests

**Pull request merged**